homeassistant/loader.py

```diff
@@ -1441,6 +1419,222 @@ async def async_get_integrations(
+            )
+        except Exception as exc:  # noqa: BLE001
+            _log_dependencies_exceptions(itg, exc)
+            return None
+
+    resolve_dependencies_tasks = {
+        itg.domain: create_eager_task(
+            task(itg),
+            name=f"resolve dependencies {itg.domain}",
+            loop=hass.loop,
+        )
+        for itg in integrations
+    }
+
+    result = await asyncio.gather(*resolve_dependencies_tasks.values())
+
+    return {
+        domain: deps
+        for domain, deps in zip(resolve_dependencies_tasks, result, strict=True)
+        if deps is not None
+    }
+
+
+async def resolve_integrations_after_dependencies(
+    hass: HomeAssistant,
+    integrations: Iterable[Integration],
+    possible_after_dependencies: set[str] | None = None,
+    *,
+    ignore_exceptions: bool = False,
+) -> dict[str, set[str]]:
+    """Resolve all dependencies, including after_dependencies, for integrations.
+
+    Detects circular dependencies and missing integrations.
+    """
+    resolved: dict[str, set[str] | Exception] = {}
+
+    async def task(itg: Integration) -> set[str] | None:
+        if ignore_exceptions:
+            return await _do_resolve_dependencies(
+                itg,
+                cache=_ResolveDependenciesCache.from_dict(resolved),
+                possible_after_dependencies=possible_after_dependencies,
+                ignore_exceptions=ignore_exceptions,
+            )
+
+        try:
+            return await _do_resolve_dependencies(
+                itg,
+                cache=_ResolveDependenciesCache.from_dict(resolved),
+                possible_after_dependencies=possible_after_dependencies,
+                ignore_exceptions=ignore_exceptions,
+            )
+        except Exception as exc:  # noqa: BLE001
+            _log_dependencies_exceptions(itg, exc)
+            return None
+
+    resolve_dependencies_tasks = {
+        itg.domain: create_eager_task(
+            task(itg),
+            name=f"resolve after dependencies {itg.domain}",
+            loop=hass.loop,
+        )
+        for itg in integrations
+    }
+
+    result = await asyncio.gather(*resolve_dependencies_tasks.values())
+
+    return {
+        domain: deps
+        for domain, deps in zip(resolve_dependencies_tasks, result, strict=True)
+        if deps is not None
+    }
+
+
+@dataclass
+class _ResolveDependenciesCache:
+    """Cache for resolve_dependencies."""
+
+    retrieve: Callable[[Integration], set[str] | Exception | None]
+    update: Callable[[Integration, set[str] | Exception], None]
+
+    @classmethod
+    def from_dict(
+        cls,
+        cache_dict: dict[str, set[str] | Exception],
+    ) -> _ResolveDependenciesCache:
+        """Create a cache from a dictionary."""
+        return cls(
+            lambda itg: cache_dict.get(itg.domain),
+            lambda itg, all_dependencies: cache_dict.__setitem__(
+                itg.domain, all_dependencies
+            ),
+        )
```

Comment:
```suggestion
        return cls(cache_dict.get, functools.partial(operator.setitem, cache_dict))
```

This looks to be reimplementing a `dict` under the hood. Is this complexity needed?